org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIDebugTarget.java

```diff
@@ -1755,7 +1756,29 @@ protected void resumeThreads() throws DebugException {
 	 */
 	public void prepareToSuspendByBreakpoint(JavaBreakpoint breakpoint) {
 		setSuspended(true);
-		suspendThreads();
+		Job job = new Job("Suspend all threads from prepareToSuspendByBreakpoint") { //$NON-NLS-1$
```

Comment:
OK, now the first question is: what if we have **N > 2** suspendAll / resumeAll jobs scheduled?
Does it make sense to run them **all** serialized? I guess no.

Note: because the execution order is **not guaranteed** by current code, and also it is impossible to guarantee it also if we would like to do so, we are free to discard all but the last job from the queue (so one is running, one is waiting to run).

Something like `org.eclipse.debug.internal.ui.model.elements.SerialExecutor` with a size bounded (max two) queue?

Ideally also this job would be managed by `EventDispatcher` so debug target can forward the operation like `getEventDispatcher().suspendAllThreads()`. This is to allow the job to be terminated / managed by `EventDispatcher`.